### PR TITLE
Add Vaihingen Segmentation

### PR DESCRIPTION
**Pull request by isaaccorley**
This PR adds the [Vaihingen](https://www2.isprs.org/commissions/comm2/wg4/benchmark/2d-sem-label-vaihingen/) dataset from the [WG II/4 2D Semantic Labeling Contest](https://www2.isprs.org/commissions/comm2/wg4/benchmark/semantic-labeling/).

Notes:
- I used the specified colormap found [here](https://www2.isprs.org/commissions/comm2/wg4/benchmark/semantic-labeling/).
- All labels are provided but I split the dataset by filenames into the train/test sets used in the challenge.
- The images are not georefenced but I'm reading them using rasterio and it complains that they aren't georeferenced in the tests.
- Images are all of different sizes
- I was able to abstract the file structure away enough that in a future PR we can make a parent class to combine the Potsdam/Vaihingen datasets or reduce copy/pasted code.

Examples:

image

### CODE REVIEW

**Comment by adamjstewart** [comment]
> The images are not georefenced but I'm reading them using rasterio and it complains that they aren't georeferenced in the tests.

I think that means we just need to read them with PIL instead?

File: tmp.ipynb [hunk 1]
```diff
@@ -0,0 +1,70 @@
+{
```
Comment (adamjstewart) [inline tmp.ipynb:1]: I assume this was added by accident?

File: torchgeo/datasets/utils.py [hunk 2]
```diff
@@ -459,3 +460,24 @@ def draw_semantic_segmentation_masks(
...
+
+
+def rgb_to_mask(  # type: ignore[type-arg]
+    rgb: np.ndarray, colors: List[Tuple[int, int, int]]
+) -> np.ndarray:  # type: ignore[type-arg]
+    """Converts an RGB colormap mask to a integer mask.
+
+    Args:
+        rgb: array mask of coded with RGB tuples
+        colors: list of RGB tuples to convert to integer indices
```
Comment (adamjstewart) [inline torchgeo/datasets/utils.py:472]: There should be an empty line between the Args and Returns sections

File: torchgeo/datasets/vaihingen.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+from torch.utils.data import DataLoader
+from torchvision.transforms import Compose
+
+from ..datasets.utils import dataset_split, draw_semantic_segmentation_masks
+from .geo import VisionDataset
+from .utils import check_integrity, extract_archive, rgb_to_mask
+
+
+class Vaihingen(VisionDataset):
+    """Vaihingen dataset.
```
Comment (adamjstewart) [inline torchgeo/datasets/vaihingen.py:24]: Same question as your other dataset, I think we'll want to make this name more specific

**Review by adamjstewart — COMMENTED** [review]
Potsdam and Vaihingen seem very similar, I wonder if they belong in a single `isprs.py` file or something like that. I guess it depends on how many datasets ISPRS has and whether there is any logical grouping between them.

**Comment by isaaccorley** [comment]
Refactored to load with PIL. Changed everywhere to Vaihingen2D.

File: tests/datasets/test_vaihingen.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+import pytest
+import torch
+import torch.nn as nn
+from _pytest.fixtures import SubRequest
+from _pytest.monkeypatch import MonkeyPatch
+
+from torchgeo.datasets import Vaihingen2D, Vaihingen2DDataModule
+
+
+class TestVaihingen:
```
Comment (adamjstewart) [inline tests/datasets/test_vaihingen.py:18]:
```suggestion
class TestVaihingen2D:
```

File: tests/datasets/test_vaihingen.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+    def test_plot(self, dataset: Vaihingen2D) -> None:
+        x = dataset[0].copy()
+        dataset.plot(x, suptitle="Test")
+        dataset.plot(x, show_titles=False)
+        x["prediction"] = x["mask"].clone()
+        dataset.plot(x)
+
+
+class TestVaihingenDataModule:
```
Comment (adamjstewart) [inline tests/datasets/test_vaihingen.py:85]:
```suggestion
class TestVaihingen2DDataModule:
```

File: torchgeo/datasets/vaihingen.py [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+from .geo import VisionDataset
+from .utils import check_integrity, extract_archive, rgb_to_mask
+
+
+class Vaihingen2D(VisionDataset):
+    """Vaihingen 2D Semantic Segmentation dataset.
+
+    The `Vaihingen <https://www2.isprs.org/commissions/comm2/wg4/benchmark/2d-sem-label-vaihingen/>`_
+    dataset is a dataset urban semantic segmentation used in the 2D Semantic Labeling
+    Contest - Vaihingen. This dataset uses the "ISPRS_semantic_labeling_Vaihingen.zip"
```
Comment (adamjstewart) [inline torchgeo/datasets/vaihingen.py:29]: Same comments as other PR, this sentence does not make sense

File: torchgeo/datasets/vaihingen.py [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+    ]
+
+    def __init__(
+        self,
+        root: str = "data",
+        split: str = "train",
+        transforms: Optional[Callable[[Dict[str, Tensor]], Dict[str, Tensor]]] = None,
+        checksum: bool = False,
+    ) -> None:
+        """Initialize a new Vaihingen dataset instance.
```
Comment (adamjstewart) [inline torchgeo/datasets/vaihingen.py:124]:
```suggestion
        """Initialize a new Vaihingen2D dataset instance.
```

Comment (adamjstewart) [reply torchgeo/datasets/vaihingen.py:124]: Or just:
```suggestion
        """Initialize a new dataset instance.
```

File: torchgeo/datasets/vaihingen.py (hunk 6 above)
Comment (isaaccorley) [reply torchgeo/datasets/vaihingen.py:29]: It should be "a dataset for urban semantic..."

File: torchgeo/datasets/vaihingen.py [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+from ..datasets.utils import dataset_split, draw_semantic_segmentation_masks
+from .geo import VisionDataset
+from .utils import check_integrity, extract_archive, rgb_to_mask
+
+
+class Vaihingen2D(VisionDataset):
+    """Vaihingen 2D Semantic Segmentation dataset.
+
+    The `Vaihingen <https://www2.isprs.org/commissions/comm2/wg4/benchmark/2d-sem-label-vaihingen/>`_
+    dataset is a dataset urban semantic segmentation used in the 2D Semantic Labeling
```
Comment (adamjstewart) [inline torchgeo/datasets/vaihingen.py:28]:
```suggestion
    dataset is a dataset for urban semantic segmentation used in the 2D Semantic Labeling
```

**Comment by calebrob6** [comment]
RIP `tmp.ipynb`

**Review by calebrob6 — APPROVED** [review]

**Review by adamjstewart — APPROVED** [review]
